File: app/controllers/flow_auth_controller.rb
class FlowAuthController < ApplicationController
  protect_from_forgery with: :null_session

  def authenticate
    raw_payload = params[:authorization_payload] || params.dig(:flow_auth, :authorization_payload)

    unless raw_payload.present?
      puts "[FlowAuth] ❌ Missing authorization payload"
      return render json: { error: "Missing authorization payload" }, status: :bad_request
    end

    payload = raw_payload.to_unsafe_h rescue raw_payload.to_h rescue nil

    unless payload.is_a?(Hash)
      puts "[FlowAuth] ❌ Payload is not a valid hash"
      return render json: { error: "Authorization payload is not a valid hash" }, status: :bad_request
    end

    flow_address = payload["addr"]
    services     = payload["services"]

    unless flow_address.present? && services.is_a?(Array)
      puts "[FlowAuth] ❌ Invalid Flow address or services"
      return render json: { error: 'Invalid Flow data' }, status: :bad_request
    end

    signature = services.find { |s| s["type"] == "user-signature" }
    unless signature
      puts "[FlowAuth] ❌ User signature service missing"
      return render json: { error: "User signature service not found" }, status: :bad_request
    end

    user = User.find_or_create_by(flow_address: flow_address)

    if user.persisted?
      if user.created_at == user.updated_at
        puts "[FlowAuth] ✅ New user created: ID=#{user.id}, Flow Address=#{user.flow_address}"
      else
        puts "[FlowAuth] 🔁 Existing user found: ID=#{user.id}, Flow Address=#{user.flow_address}"
      end
    else
      puts "[FlowAuth] ❌ User not persisted (unexpected)"
    end

    session[:user_id] = user.id
    puts "[FlowAuth] 💾 Session set for user ID=#{user.id}"

    render json: {
      message: "Authentication successful",
      redirect_to: Rails.application.routes.url_helpers.edit_user_path(user)
    }, status: :ok
    
  rescue => e
    Rails.logger.error("Flow auth error: #{e.class.name} - #{e.message}")
    puts "[FlowAuth] ❌ Exception raised: #{e.class} - #{e.message}"
    render json: { error: "Internal Server Error" }, status: :internal_server_error
  end


  def logout
    reset_session
    cookies.delete(:auth_token)
    render json: { message: 'Logout successful' }, status: :ok
  rescue => e
    Rails.logger.error("Logout error: #{e.class.name} - #{e.message}")
    render json: { error: 'Internal Server Error' }, status: :internal_server_error
  end
end

File: app/javascript/controllers/flow_auth_controller.js
import { Controller } from "@hotwired/stimulus"
import * as fcl from '@onflow/fcl'
import { init } from '@onflow/fcl-wc'

export default class extends Controller {
  static targets = ["loginButton", "logoutButton"]

  connect() {
    fcl.config({
      "flow.network": "testnet",
      "accessNode.api": "https://rest-testnet.onflow.org",
      "discovery.wallet": "https://fcl-discovery.onflow.org/testnet/authn",
      "walletconnect.projectId": "11069f4303b90a2b1628ea2c1e4cf861",
      "app.detail.title": "ViralTrust",
      "app.detail.icon": "https://assets.reown.com/reown-profile-pic.png",
      "app.detail.description": "Join ViralTrust.",
      "app.detail.url": "https://viral-trust.com"
    })

    this.updateUserStatus(false)
  }

  async login() {
    try {
      const user = await fcl.authenticate()
      const walletAddress = user.addr

      const hasSignature = user.services.some(service => service.type === "user-signature")
      if (!hasSignature) {
        alert("User signature service not found.")
        return
      }

      const response = await fetch('/flow_auth/authenticate', {
        method: 'POST',
        headers: {
          "Content-Type": "application/json",
          "X-CSRF-Token": document.querySelector('meta[name="csrf-token"]').content,
        },
        body: JSON.stringify({ authorization_payload: user }),
      })

      const data = await response.json()
      if (response.ok && data.message === "Authentication successful") {
        this.updateUserStatus(true, walletAddress)
      } else {
        alert(data.error || "Authentication failed")
      }
    } catch (error) {
      console.error("Flow login error:", error)
      alert("Une erreur est survenue pendant l’authentification.")
    }
  }

  async logout() {
    try {
      localStorage.clear()

      const response = await fetch('/flow_auth/logout', {
        method: 'DELETE',
        headers: {
          "Content-Type": "application/json",
          "X-CSRF-Token": document.querySelector('meta[name="csrf-token"]').content,
        },
      })

      if (response.ok) {
        this.updateUserStatus(false)
        window.location.href = '/pages/home'
      } else {
        const error = await response.json()
        alert(error.error || "Logout failed")
      }
    } catch (err) {
      console.error("Logout error:", err)
      alert("Une erreur est survenue pendant la déconnexion.")
    }
  }

  updateUserStatus(connected, address = null) {
    if (this.hasLoginButtonTarget)
      this.loginButtonTarget.style.display = connected ? "none" : "block"

    if (this.hasLogoutButtonTarget)
      this.logoutButtonTarget.style.display = connected ? "block" : "none"
  }
}
